src/endurance-test/src/main.rs

```diff
@@ -0,0 +1,230 @@
+    retry_policy::{Aip194Strict, RetryPolicyExt},
+};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::{Duration, Instant};
+
+const TOO_MANY_ERRORS: u64 = 100;
+static ERROR_COUNT: AtomicU64 = AtomicU64::new(0);
+static SUCCESS_COUNT: AtomicU64 = AtomicU64::new(0);
+static UPDATE_COUNT: AtomicU64 = AtomicU64::new(0);
+
+#[tokio::main(flavor = "multi_thread", worker_threads = 10)]
+async fn main() {
+    if let Err(e) = start_workers().await {
+        report_error(e, "main");
+    }
+}
+
+/// Continuously perform RPCs to a given secret.
+async fn worker(
+    client: sm::client::SecretManagerService,
+    secret: String,
+    task_id: usize,
+    total_workers: usize,
+) -> anyhow::Result<()> {
+    // First create a secret version, to ensure the loop will succeed.
+    let mut version = update_secret(&client, &secret).await?;
+
+    // We want to create a new version every minute on average. That will
+    // keep this test well below the quota:
+    //   https://cloud.google.com/secret-manager/quotas
+    let add_version_period = Duration::from_secs(60).mul_f32(total_workers as f32);
+    let report_period = Duration::from_secs(10);
+
+    let mut last_add_version = Instant::now();
+    let mut last_report = Instant::now();
+    let mut error_count = 0;
+    let mut success_count = 0;
+    let mut update_count = 0;
+    loop {
+        if Instant::now() >= last_add_version + add_version_period {
+            version = update_secret(&client, &secret).await?;
+            last_add_version = Instant::now();
+            update_count += 1;
+        }
+        if Instant::now() >= last_report + report_period {
+            SUCCESS_COUNT.fetch_add(success_count, Ordering::SeqCst);
+            ERROR_COUNT.fetch_add(error_count, Ordering::SeqCst);
+            UPDATE_COUNT.fetch_add(update_count, Ordering::SeqCst);
+            if task_id == 0 {
+                report_info(
+                    format!(
+                        "success_count={}, error_count={}, version_count={}, current_success_count={}, current_error_count={}, total_workers={}",
+                        SUCCESS_COUNT.load(Ordering::Relaxed),
+                        ERROR_COUNT.load(Ordering::Relaxed),
+                        UPDATE_COUNT.load(Ordering::Relaxed),
+                        success_count,
+                        error_count,
+                        total_workers
+                    ),
+                    "task[0]",
+                );
+            }
+            last_report = Instant::now();
+            error_count = 0;
+            success_count = 0;
+        }
+        // We want to average about 30,000 requests per minute. That will keep the
+        // test well below the quota:
+        //   https://cloud.google.com/secret-manager/quotas
+        let wait = tokio::time::Instant::now()
+            + Duration::from_secs(60)
+                .mul_f32(total_workers as f32)
+                .div_f32(90_000_f32);
```

Comment:
Why is this 3x the rate cited in the comment

Reply:
Did not update the comments on numbers after testing.